File: app/src/main/java/com/veganbeauty/app/features/routine/RoutineAlarmScheduler.java
package com.veganbeauty.app.features.routine;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import com.veganbeauty.app.data.local.ProfileSession;
import com.veganbeauty.app.features.account.notification.NotificationScheduleHelper;

import java.util.Calendar;

public class RoutineAlarmScheduler {

    private static final String TAG = "RoutineAlarmScheduler";
    private static final int REQUEST_MORNING_MAIN = 1001;
    private static final int REQUEST_EVENING_MAIN = 1002;
    private static final int REQUEST_MORNING_LEAD = 1003;
    private static final int REQUEST_EVENING_LEAD = 1004;

    public static void scheduleRoutineAlarm(Context context, String type, int hour, int minute, boolean isLead) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, hour);
        calendar.set(Calendar.MINUTE, minute);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        if (calendar.getTimeInMillis() <= System.currentTimeMillis()) {
            calendar.add(Calendar.DAY_OF_YEAR, 1);
        }

        Bundle extras = new Bundle();
        extras.putString("REMINDER_TYPE", type);
        extras.putBoolean("IS_LEAD_REMINDER", isLead);

        NotificationScheduleHelper.scheduleBroadcastAlarm(
                context,
                requestCodeFor(type, isLead),
                RoutineAlarmReceiver.class,
                extras,
                calendar.getTimeInMillis(),
                "SKIN_REMINDER"
        );
        Log.d(TAG, "Scheduled " + type + (isLead ? " lead" : " main")
                + " at " + hour + ":" + String.format("%02d", minute)
                + " next=" + calendar.getTime());
    }

    public static void cancelRoutineAlarm(Context context, String type) {
        cancelSingleAlarm(context, type, false);
        cancelSingleAlarm(context, type, true);
    }

    private static void cancelSingleAlarm(Context context, String type, boolean isLead) {
        NotificationScheduleHelper.cancelBroadcastAlarm(
                context, requestCodeFor(type, isLead), RoutineAlarmReceiver.class);
    }

    private static int requestCodeFor(String type, boolean isLead) {
        if ("MORNING".equals(type)) {
            return isLead ? REQUEST_MORNING_LEAD : REQUEST_MORNING_MAIN;
        }
        return isLead ? REQUEST_EVENING_LEAD : REQUEST_EVENING_MAIN;
    }

    public static void rescheduleAlarms(Context context) {
        if (!ProfileSession.isNotiEnabled(context)) {
            cancelRoutineAlarm(context, "MORNING");
            cancelRoutineAlarm(context, "EVENING");
            return;
        }

        NotificationScheduleHelper.remindExactAlarmIfNeeded(context);

        boolean morningEnabled = ProfileSession.isMorningReminderEnabled(context);
        boolean eveningEnabled = ProfileSession.isEveningReminderEnabled(context);
        boolean leadEnabled = ProfileSession.isLeadReminderEnabled(context);

        if (morningEnabled) {
            int[] morning = parseTime(ProfileSession.getMorningReminderTime(context), 6, 30);
            scheduleRoutineAlarm(context, "MORNING", morning[0], morning[1], false);
            if (leadEnabled) {
                int[] lead = subtractMinutes(morning[0], morning[1], 15);
                scheduleRoutineAlarm(context, "MORNING", lead[0], lead[1], true);
            } else {
                cancelSingleAlarm(context, "MORNING", true);
            }
        } else {
            cancelRoutineAlarm(context, "MORNING");
        }

        if (eveningEnabled) {
            int[] evening = parseTime(ProfileSession.getEveningReminderTime(context), 21, 45);
            scheduleRoutineAlarm(context, "EVENING", evening[0], evening[1], false);
            if (leadEnabled) {
                int[] lead = subtractMinutes(evening[0], evening[1], 15);
                scheduleRoutineAlarm(context, "EVENING", lead[0], lead[1], true);
            } else {
                cancelSingleAlarm(context, "EVENING", true);
            }
        } else {
            cancelRoutineAlarm(context, "EVENING");
        }
    }

    private static int[] parseTime(String time, int defaultHour, int defaultMinute) {
        int hour = defaultHour;
        int minute = defaultMinute;
        if (time != null) {
            String[] parts = time.split(":");
            try {
                if (parts.length > 0) hour = Integer.parseInt(parts[0].trim());
                if (parts.length > 1) minute = Integer.parseInt(parts[1].trim());
            } catch (NumberFormatException ignored) {
            }
        }
        return new int[]{hour, minute};
    }

    private static int[] subtractMinutes(int hour, int minute, int delta) {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, hour);
        cal.set(Calendar.MINUTE, minute);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        cal.add(Calendar.MINUTE, -delta);
        return new int[]{cal.get(Calendar.HOUR_OF_DAY), cal.get(Calendar.MINUTE)};
    }
}

File: app/src/main/java/com/veganbeauty/app/features/routine/RoutineAlarmScheduler.kt
package com.veganbeauty.app.features.routine

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import java.util.Calendar

object RoutineAlarmScheduler {
    fun scheduleRoutineAlarm(context: Context, type: String, hour: Int, minute: Int, isLead: Boolean) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, RoutineAlarmReceiver::class.java).apply {
            putExtra("REMINDER_TYPE", type)
            putExtra("IS_LEAD_REMINDER", isLead)
        }
        val requestCode = if (type == "MORNING") 1001 else 1002
        val pendingIntent = PendingIntent.getBroadcast(
            context,
            requestCode,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val calendar = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, hour)
            set(Calendar.MINUTE, minute)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
            if (timeInMillis < System.currentTimeMillis()) {
                add(Calendar.DAY_OF_YEAR, 1)
            }
        }

        try {
            alarmManager.setExactAndAllowWhileIdle(
                AlarmManager.RTC_WAKEUP,
                calendar.timeInMillis,
                pendingIntent
            )
        } catch (e: SecurityException) {
            // Fallback for Android 12+ if SCHEDULE_EXACT_ALARM permission isn't granted yet
            alarmManager.set(
                AlarmManager.RTC_WAKEUP,
                calendar.timeInMillis,
                pendingIntent
            )
        }
    }

    fun cancelRoutineAlarm(context: Context, type: String) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(context, RoutineAlarmReceiver::class.java)
        val requestCode = if (type == "MORNING") 1001 else 1002
        val pendingIntent = PendingIntent.getBroadcast(
            context,
            requestCode,
            intent,
            PendingIntent.FLAG_NO_CREATE or PendingIntent.FLAG_IMMUTABLE
        )
        if (pendingIntent != null) {
            alarmManager.cancel(pendingIntent)
            pendingIntent.cancel()
        }
    }

    fun rescheduleAlarms(context: Context) {
        val morningEnabled = com.veganbeauty.app.data.local.ProfileSession.isMorningReminderEnabled(context)
        val eveningEnabled = com.veganbeauty.app.data.local.ProfileSession.isEveningReminderEnabled(context)
        val leadEnabled = com.veganbeauty.app.data.local.ProfileSession.isLeadReminderEnabled(context)

        if (morningEnabled) {
            val morningTime = com.veganbeauty.app.data.local.ProfileSession.getMorningReminderTime(context)
            val morningParts = morningTime.split(":")
            var mHour = morningParts.getOrNull(0)?.toIntOrNull() ?: 6
            var mMinute = morningParts.getOrNull(1)?.toIntOrNull() ?: 30
            if (leadEnabled) {
                val cal = Calendar.getInstance().apply {
                    set(Calendar.HOUR_OF_DAY, mHour)
                    set(Calendar.MINUTE, mMinute)
                }
                cal.add(Calendar.MINUTE, -15)
                mHour = cal.get(Calendar.HOUR_OF_DAY)
                mMinute = cal.get(Calendar.MINUTE)
            }
            scheduleRoutineAlarm(context, "MORNING", mHour, mMinute, leadEnabled)
        } else {
            cancelRoutineAlarm(context, "MORNING")
        }

        if (eveningEnabled) {
            val eveningTime = com.veganbeauty.app.data.local.ProfileSession.getEveningReminderTime(context)
            val eveningParts = eveningTime.split(":")
            var eHour = eveningParts.getOrNull(0)?.toIntOrNull() ?: 21
            var eMinute = eveningParts.getOrNull(1)?.toIntOrNull() ?: 45
            if (leadEnabled) {
                val cal = Calendar.getInstance().apply {
                    set(Calendar.HOUR_OF_DAY, eHour)
                    set(Calendar.MINUTE, eMinute)
                }
                cal.add(Calendar.MINUTE, -15)
                eHour = cal.get(Calendar.HOUR_OF_DAY)
                eMinute = cal.get(Calendar.MINUTE)
            }
            scheduleRoutineAlarm(context, "EVENING", eHour, eMinute, leadEnabled)
        } else {
            cancelRoutineAlarm(context, "EVENING")
        }
    }
}
